src/server/routes/entity/entity.js

```diff
@@ -61,9 +61,86 @@ const entityComponents = {
 	work: WorkPage
 };
 
+/**
+ * Creates a new database record whenever an editor visits an entity, to allow
+ * tracking of popular entities and handling of the page visit achievement.
+ *
+ * @param {Object} orm An initialized bookbrainz-data ORM
+ * @param {string} bbid The BBID of the entity that was visited
+ * @param {?number} editorID The ID of the editor, if logged in, otherwise null
+ *
+ * @returns {Promise<Object|boolean>} An object containing achievement data, or
+ *          false if an error occured
+ */
+function logEntityVisit(orm: Object, bbid: string, editorID: ?number) {
+	const {EditorEntityVisits} = orm;
+
+	if (!editorID) {
+		return Promise.resolve(false);
+	}
+
+	return new EditorEntityVisits({bbid, editorId: editorID})
+		.save(null, {method: 'insert'})
+		.then(() => achievement.processPageVisit(orm, editorID))
+		.catch(
+			// error caused by duplicates we do not want in database
+			() => Promise.resolve(false)
+		);
+}
+
+/**
+ * Takes a list of achievement unlock IDs and fetches achievement information
+ * for each. Then builds a list of achievement information represented by these
+ * unlock IDs to be used in the unlock alert displayed to the editor.
+ *
+ * @param {Object} orm An initialized bookbrainz-data ORM
+ * @param {number[]} alertIDs A list of achievement unlock IDs to process
+ * @param {?number} editorID The ID of the editor, if logged in, otherwise null
+ *
+ * @returns {Promise<Object[]|boolean>} A list containing achievement
+ *          information, or false if an error occured
+ */
+function processAlertIDs(orm: Object, alertIDs: number[], editorID: ?number) {
```

Comment:
Is there a reason why editorID is optional, considering the if block at lines 109-111 ?

Reply:
This is because the `displayEntity` function calls `processAlertIDs`, and there isn't always a logged-in editor when `displayEntity` is called. We could move the condition into `displayEntity` - wouldn't be difficult. Previously this check was on line 78.

Reply:
Yes, that makes perfect sense.
I think I was just tired when I wrote this :)